docs/proposals/scheduling/20220701-schedule-gang.md

```diff
@@ -0,0 +1,341 @@
+
+1. Firstly, compare the priorities of the two pods, the higher priority is at the front of the queue.
+
+2. Secondly, compare create-time-stamp of two pods, if pod belongs to a gang, then we compare create-time-stamp of the gang, 
+the one created first will be at the front of the queue.
+
+3. Finally, compare pod's namespace, if pod belongs to a gang, then we compare gang name. 
+
+```go
+type QueueSortPlugin interface{
+    QueueSort(*QueuedPodInfo, *QueuedPodInfo) bool
+}
+```
+
+#### GangSchedulingPlugin
+##### Data-Structure
+
+###### strict-mode and non-strict-mode
+As mentioned above, in `strict-mode`, if a pod failed to be scheduled, the pods that have been successfully scheduled in 
+this scheduling cycle will be rolled back, and the remaining pods waiting for scheduling will be rejected in 
+PreFilter check util this scheduling cycle passed. We call this mode is `strict-mode`.
+
+In `non-strict-mode`, if a pod failed to be scheduled, it has no impact on any other pod. We will continue to accumulate 
+the allocated pod until the condition of gang is met. This process is friendly to gangs with large number of pods, but it 
+will increase the risk of resource deadlock between gangs. For example, the quota of the quota group is 10(quota will be proposed later), 
+and the user submits three gangs with 5 pods. Due to various plugin constraints, gang1\2\3 may allocate resources of 3\3\4 respectively. 
+Since the quota group's quota is full, there will be no new resource scheduling. We call this is resource deadlock of resource gang.
+In future proposal, we will try to fix this problem.
+
+###### Gang
+
+We design the gang to record gang status in scheduler memory, it has the `Bundles` field to store the gang's children by bundle name.
+We can also find BundleInfo from `PodToBundleMap` field according to the pod's NamespacedName. We can check the `ResourceSatisfied`
+field to see if the gang is already has the minimum number assumed pods in each bundle. 
+
+It should be noted that, if the resource accumulation conditions of gang are met, then some pods failed in the process of binding;
+Or some bound pods are preempted\rescheduled, should the constraints of gang still be effective in the process of resource reallocation? 
+Because the initial purpose of gang is to require pods to be pulled up at the same time, if some pods have been pulled up, 
+then the subsequent gang behavior is meaningless. Therefore, when `resourcequalified=true`, all subsequent resource allocations 
+are no longer constrained by gang rules, and their performance is similar to ordinary pod.
+
+As mentioned above, `WaitTime` is the max wait time since first pod comes to permit stage. if `WaitTime` is timeout, 
+scheduler will roll back all assumed pods, update each pod's annotation with `koordinater.io.gang/timeout=true`, and
+won't schedule these pods anymore. User should pay attention to this status and delete pods timely.
```

Comment:
"User should pay attention to this status and delete pods timely", but `We design the gang to record gang status in scheduler memory `. So is the gang status just in memory? so how to the user pay attention to the status? It's confusing here.

Reply:
through monitor pod's annotation of `koordinater.io.gang/timeout=true`,  we just want to have a way to notify user gang is timeout.